Add generic support for all fluid + barrel types

macintacos opened the pull request:
Added a bunch of fluids that were missing for some mods that I'm using in a new playthrough. Tested locally and it seems to work great.

Notably: I didn't spend any time looking into how to do this in a generic way for any type of modded fluid. I've never really messed with Factorio mods before, so I figured that would be a bit out of my league.

macintacos committed:
feat: add support for Maraxsis fluids

macintacos committed:
feat: add support for Corrundum fluids

macintacos committed:
feat: add support for Cerys fluids

macintacos committed:
feat; add support for On Wayward Seas fluids

macintacos committed:
feat: add support for Shattered Planet fluids

macintacos committed:
chore: bump version, update changelog

macintacos commented:
I'm actually experimenting with a generic way of doing this, because my curiosity got the best of me. If I don't figure something out within a couple of hours, I'll comment here.

macintacos commented:
I've given up for now 😅 it feels like this is possible though. The trickiest part is actually figuring out what the _first_ technology that unlocks a given fluid. In the details for the various fluid recipe prototypes, it seems like there are multiple `unlock-recipe`'s designated even if they're not the very first tech to unlock it. I would imagine we'd need to build some kind of small "solver" to figure it out, walking up the tech tree until we get to "fluid-handling" (or something like that) and keeping track of which one has the shortest path, and designating that as its true "unlock recipe". I'm doing a bit too much bumbling around in the API right now and I kinda want to get back to playing, so I'll probably pick this back up at another time.

@T3sT3ro Let me know if you'd like to just accept this PR, and maybe revisit the "generic" solution in another PR.

macintacos commented:
Lmao okay, I shouldn't have jumped the gun - I think I got something working! Will push a PR up and adjust the name of this PR to detail it. Might adjust the version to 2.0, since it is a decent change to functionality. Thoughts on that?

T3sT3ro commented:
Thank for PR, I'm only looking into it right now because I didn't remember how my dev setup worked and couldn't find the time earlier. I'll try to merge it in today or tomorrow. I have a question though: what does https://mods.factorio.com/mod/BarrelFluidIcons mod do? I tried enabling and disabling it and reading the description several times, but still can't find the difference or see what's changed.

As for the hardcoded technologies vs generic way — I think hardcoded is easier to implement in this case, but if you have a generic way to walk up the graph which wouldn't require explicitly adding compatibility for each mod it would be very nice and I would be happy to merge it :) (not sure how much time I'll have though in the following days so sorry if it won't be immediate ;)

macintacos committed:
feat!: dynamically parse fluid techs to get barrel unlock techs

macintacos committed:
chore: bump major version

macintacos commented:
Oh that mod simply adds barrel recipes for _all_ fluids including modded fluids, if it can find them. I started a new playthrough with a ton of new planets, and many of them add their own fluids, so without this mod's functionality I was always presented with a list of like 30 fluid barrels that I hadn't even unlocked yet 😅 

I just pushed up the generic implementation! It seems to work perfectly great on my machine, but it's entirely possible that I missed something. Please let me know if you have any questions. No worries about the timeline - I have it manually installed locally anyway, so I'm fine with waiting lol.

macintacos committed:
fix: formatting

T3sT3ro commented on src/data-final-fixes.lua:
```diff
@@ -1,49 +1,105 @@
-  ['fluoroketone-hot-barrel'] = 'cryogenic-plant',
-  ['empty-fluoroketone-hot-barrel'] = 'cryogenic-plant',
-}
+-- Helper function to convert a table to a string
+---@param tableToDump table - The table that you want to conver to a string
+---@return string - The table, dumped as an ugly string
+local function dumpTable(tableToDump)
+  if type(tableToDump) == "table" then
+    local s = "{ "
+    for k, v in pairs(tableToDump) do
+      if type(k) ~= "number" then
+        k = '"' .. k .. '"'
+      end
+      s = s .. "[" .. k .. "] = " .. dumpTable(v) .. ","
+    end
+    return s .. "} "
+  else
+    return tostring(tableToDump)
+  end
+end
```
I reckon something like that exists: https://lua-api.factorio.com/latest/classes/LuaHelpers.html#table_to_json

T3sT3ro commented:
Something is not right:
- water [un]barreling recipes are unlocked by "space platform thruster" instead of "fluid handling"
- crude oil [un]barreling cannot be unlocked anymore (should be unlocked with "oil gathering")

More issues probably occur, I did not check further for now.

macintacos committed:
fix: use `helpers.table_to_json`

T3sT3ro commented:
Maybe the assumption that all fluids are products of recipes doesn't cover all edge cases like with the crude oil which is mined by pumpjacks. For that case I am leaning towards a hardcoded solution where certain recipes require manual assignment, which is more straightforward and explicit. Generic one requires more testing and consideration for edge cases.

macintacos committed:
fix: add ability to skip and explicitly override recpies

macintacos commented:
Ah - up to you @T3sT3ro. I just pushed a fix that handles some edge cases like water and crude oil. I manually went through all of the fluids that had barrel recipes and added some manual overrides, and explicitly skipped processing water since there isn't even really a tech associated with it from what I can tell.

Everything seems to be working good now, but let me know if you'd like me to just switch back to the hard-coded approach. I might keep this fork for my own purposes and keep iterating if that's the case 😅

macintacos committed:
fix: remove link to other mod in changelog (in retrospect, doesn't make sense to have it there)

T3sT3ro committed:
Generate vscode launch configuration for debugging

T3sT3ro committed:
Refactor for 2.0

- logs are now written to general factorio log
- exceptions are handled uniformly
- used existing utils for debug logging data (serpent lib)

T3sT3ro committed:
Move Maraxsis saline water to "planet-discovery-maraxsis"

T3sT3ro committed:
Update mod info, add kudos to @macintacos

T3sT3ro committed:
Easter egg

T3sT3ro reviewed (approved):
Great work! Tested and working, I did some final refactor and simplification. Also I remembered this: https://lua-api.factorio.com/latest/auxiliary/libraries.html
